src/target/riscv/riscv-013.c

```diff
@@ -731,29 +516,52 @@ static uint32_t riscv013_get_dmi_address(const struct target *target, uint32_t a
 	return address + base;
 }
 
+static int batch_run_timeout(struct target *target, struct riscv_batch *batch);
+
+static int dmi_read(struct target *target, uint32_t *value, uint32_t address)
+{
+	struct riscv_batch *batch = riscv_batch_alloc(target, 1);
+	riscv_batch_add_dmi_read(batch, address, RISCV_DELAY_BASE);
+	int res = batch_run_timeout(target, batch);
+	if (res == ERROR_OK && value)
+		*value = riscv_batch_get_dmi_read_data(batch, 0);
+	riscv_batch_free(batch);
+	return res;
+}
+
 static int dm_read(struct target *target, uint32_t *value, uint32_t address)
 {
-	dm013_info_t *dm = get_dm(target);
-	if (!dm)
-		return ERROR_FAIL;
-	return dmi_read(target, value, address + dm->base);
+	return dmi_read(target, value, riscv013_get_dmi_address(target, address));
 }
 
 static int dm_read_exec(struct target *target, uint32_t *value, uint32_t address)
 {
 	dm013_info_t *dm = get_dm(target);
 	if (!dm)
 		return ERROR_FAIL;
+	struct riscv_batch *batch = riscv_batch_alloc(target, 1);
+	riscv_batch_add_dm_read(batch, address, RISCV_DELAY_ABSTRACT_COMMAND);
 	dm->abstract_cmd_maybe_busy = true;
-	return dmi_read_exec(target, value, address + dm->base);
+	int res = batch_run_timeout(target, batch);
+	if (res == ERROR_OK && value)
+		*value = riscv_batch_get_dmi_read_data(batch, 0);
+	riscv_batch_free(batch);
+	return res;
+}
```

Comment:
I think this functionality (`dmi_read()` and `dm_read_exec()`) could further be unified by a function like `dmi_read_batch_delay(target, value, address, delay)`

What do you think?

Reply:
The issue is with `dm->abstract_cmd_maybe_busy = true;`. 
1. It can be either left in `dm_read_exec()`, which would create a possibility of where is is left high after an allocation error and there is no abstract command running on the target right now. This is not a big deal -- assuming an abstract commnad is running shouldn't break anything, but this is a possible misrepresentation of target's state that can be avoided. Moreover, the whole `get_dm()` related part would be duplicated so the diff will increase.
2. Alternatively, `dm->abstract_cmd_maybe_busy = true;` can be moved into the proposed `dmi_read_batch_delay()`. Hovewer, it would be placed there under an `if` on delay type. I believe this would not make the code simpler.

Reply:
I don't mind the separate functions as it is now. It looks clear enough to me.

Alternatively, `dmi_read_exec()` could be created so that the call graph has the same structure in both cases and is very easy to comprehend for code readers:

- `dm_read_exec()` -> `dmi_read_exec()` -> batch functions
- `dm_read()` -> `dmi_read()` -> batch functions

But that's really a detail (optional suggestion).

Reply:
I'd like to leave it as-is.

Regarding `dmi_read_exec()` -- I can't imagine a scenario where it would be useful, outside `dm_read_exec()`, so would prefer not to add it.

Reply:
Ok, let's leave it as is.